File: nim_core_v2/ios/Classes/services/FLTConversationGroupService.swift
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import Foundation
import NIMSDK

enum V2ConversationGroupAPIType: String {
  case createConversationGroup
  case deleteConversationGroup
  case updateConversationGroup
  case addConversationsToGroup
  case removeConversationsFromGroup
  case getConversationGroup
  case getConversationGroupList
  case getConversationGroupListByIds
}

@objcMembers
class FLTConversationGroupService: FLTBaseService, FLTService {
  private static let className = "FLTConversationGroupService"

  override func onInitialized() {
    NIMSDK.shared().v2ConversationGroupService.add(self)
  }

  deinit {
    NIMSDK.shared().v2ConversationGroupService.remove(self)
  }

  func serviceName() -> String {
    ServiceType.ConversationGroupService.rawValue
  }

  func register(_ nimCore: NimCore) {
    self.nimCore = nimCore
    nimCore.addService(self)
  }

  func onMethodCalled(_ method: String, _ arguments: [String: Any], _ resultCallback: ResultCallback) {
    switch method {
    case V2ConversationGroupAPIType.createConversationGroup.rawValue:
      createConversationGroup(arguments, resultCallback)
    case V2ConversationGroupAPIType.deleteConversationGroup.rawValue:
      deleteConversationGroup(arguments, resultCallback)
    case V2ConversationGroupAPIType.updateConversationGroup.rawValue:
      updateConversationGroup(arguments, resultCallback)
    case V2ConversationGroupAPIType.addConversationsToGroup.rawValue:
      addConversationsToGroup(arguments, resultCallback)
    case V2ConversationGroupAPIType.removeConversationsFromGroup.rawValue:
      removeConversationsFromGroup(arguments, resultCallback)
    case V2ConversationGroupAPIType.getConversationGroupList.rawValue:
      getConversationGroupList(arguments, resultCallback)
    case V2ConversationGroupAPIType.getConversationGroup.rawValue:
      getConversationGroup(arguments, resultCallback)
    case V2ConversationGroupAPIType.getConversationGroupListByIds.rawValue:
      getConversationGroupListByIds(arguments, resultCallback)
    default:
      resultCallback.notImplemented()
    }
  }

  // MARK: - SDK API

  /**
   *  创建会话分组
   *
   *  @param name 分组名
   *  @param serverExtension 服务器扩展字段
   *  @param conversationIds 会话id列表
   */
  func createConversationGroup(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let name = arguments["name"] as? String else {
      parameterError(resultCallback)
      return
    }

    let serverExtension = arguments["serverExtension"] as? String
    let conversationIds = arguments["conversationIds"] as? [String]

    weak var weakSelf = self
    NIMSDK.shared().v2ConversationGroupService.createConversationGroup(name, serverExtension: serverExtension, conversationIds: conversationIds) { result in
      weakSelf?.successCallBack(resultCallback, result.toDic())
    } failure: { error in
      weakSelf?.errorCallBack(resultCallback, error.nserror.localizedDescription, Int(error.code))
      FLTALog.errorLog(FLTConversationGroupService.className, desc: "createConversationGroup error \(error.nserror.localizedDescription)")
    }
  }

  /**
   *  删除会话分组
   *
   *  @param groupId 分组id
   */
  func deleteConversationGroup(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let groupId = arguments["groupId"] as? String else {
      parameterError(resultCallback)
      return
    }

    weak var weakSelf = self
    NIMSDK.shared().v2ConversationGroupService.deleteConversationGroup(groupId) {
      weakSelf?.successCallBack(resultCallback, nil)
    } failure: { error in
      weakSelf?.errorCallBack(resultCallback, error.nserror.localizedDescription, Int(error.code))
      FLTALog.errorLog(FLTConversationGroupService.className, desc: "deleteConversationGroup error \(error.nserror.localizedDescription)")
    }
  }

  /**
   *  更新会话分组
   *
   *  @param groupId 分组id
   *  @param name 分组名
   *  @param serverExtension 服务器扩展字段
   */
  func updateConversationGroup(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let groupId = arguments["groupId"] as? String else {
      parameterError(resultCallback)
      return
    }

    weak var weakSelf = self
    let name = arguments["name"] as? String
    let serverExtension = arguments["serverExtension"] as? String
    NIMSDK.shared().v2ConversationGroupService.updateConversationGroup(groupId, name: name, serverExtension: serverExtension) {
      weakSelf?.successCallBack(resultCallback, nil)
    } failure: { error in
      weakSelf?.errorCallBack(resultCallback, error.nserror.localizedDescription, Int(error.code))
      FLTALog.errorLog(FLTConversationGroupService.className, desc: "updateConversationGroup error \(error.nserror.localizedDescription)")
    }
  }

  /**
   *  添加会话到分组
   *
   *  @param groupId 分组id
   *  @param conversationIds 会话id列表
   */
  func addConversationsToGroup(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let groupId = arguments["groupId"] as? String,
          let conversationIds = arguments["conversationIds"] as? [String] else {
      parameterError(resultCallback)
      return
    }

    weak var weakSelf = self
    NIMSDK.shared().v2ConversationGroupService.addConversations(toGroup: groupId, conversationIds: conversationIds) { results in
      let resultsDic = results.map { $0.toDic() }
      weakSelf?.successCallBack(resultCallback, ["conversationOperationResults": resultsDic])
    } failure: { error in
      weakSelf?.errorCallBack(resultCallback, error.nserror.localizedDescription, Int(error.code))
      FLTALog.errorLog(FLTConversationGroupService.className, desc: "addConversationsToGroup error \(error.nserror.localizedDescription)")
    }
  }

  /**
   *  从会话分组移除会话
   *
   *  @param groupId 分组id
   *  @param conversationIds 会话id列表
   */
  func removeConversationsFromGroup(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let groupId = arguments["groupId"] as? String,
          let conversationIds = arguments["conversationIds"] as? [String] else {
      parameterError(resultCallback)
      return
    }

    weak var weakSelf = self
    NIMSDK.shared().v2ConversationGroupService.removeConversations(fromGroup: groupId, conversationIds: conversationIds) { results in
      let resultsDic = results.map { $0.toDic() }
      weakSelf?.successCallBack(resultCallback, ["conversationOperationResults": resultsDic])
    } failure: { error in
      weakSelf?.errorCallBack(resultCallback, error.nserror.localizedDescription, Int(error.code))
      FLTALog.errorLog(FLTConversationGroupService.className, desc: "removeConversationsFromGroup error \(error.nserror.localizedDescription)")
    }
  }

  /**
   *  获取会话分组
   *
   *  @param groupId 分组Id
   */
  func getConversationGroup(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let groupId = arguments["groupId"] as? String else {
      parameterError(resultCallback)
      return
    }

    weak var weakSelf = self
    NIMSDK.shared().v2ConversationGroupService.getConversationGroup(groupId) { group in
      weakSelf?.successCallBack(resultCallback, group.toDic())
    } failure: { error in
      weakSelf?.errorCallBack(resultCallback, error.nserror.localizedDescription, Int(error.code))
      FLTALog.errorLog(FLTConversationGroupService.className, desc: "getConversationGroup error \(error.nserror.localizedDescription)")
    }
  }

  /**
   *  获取会话分组列表
   *
   */
  func getConversationGroupList(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    weak var weakSelf = self
    NIMSDK.shared().v2ConversationGroupService.getConversationGroupList { groups in
      let groupsDic = groups.map { $0.toDic() }
      weakSelf?.successCallBack(resultCallback, ["conversationGroups": groupsDic])
    } failure: { error in
      weakSelf?.errorCallBack(resultCallback, error.nserror.localizedDescription, Int(error.code))
      FLTALog.errorLog(FLTConversationGroupService.className, desc: "getConversationGroupList error \(error.nserror.localizedDescription)")
    }
  }

  /**
   *  获取会话分组列表
   *
   *  @param groupIds 分组Id列表
   */
  func getConversationGroupListByIds(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let groupIds = arguments["groupIds"] as? [String] else {
      parameterError(resultCallback)
      return
    }

    weak var weakSelf = self
    NIMSDK.shared().v2ConversationGroupService.getConversationGroupList(byIds: groupIds) { groups in
      let groupsDic = groups.map { $0.toDic() }
      weakSelf?.successCallBack(resultCallback, ["conversationGroups": groupsDic])
    } failure: { error in
      weakSelf?.errorCallBack(resultCallback, error.nserror.localizedDescription, Int(error.code))
      FLTALog.errorLog(FLTConversationGroupService.className, desc: "getConversationGroupListByIds error \(error.nserror.localizedDescription)")
    }
  }
}

// MARK: - V2NIMConversationGroupListener

extension FLTConversationGroupService: V2NIMConversationGroupListener {
  /**
   *  会话分组创建回调
   *
   *  @param conversationGroup 会话分组
   */
  func onConversationGroupCreated(_ conversationGroup: V2NIMConversationGroup) {
    notifyEvent(serviceName(), "onConversationGroupCreated", conversationGroup.toDic())
  }

  /**
   *  会话分组删除回调
   *
   *  @param groupId 分组id
   */
  func onConversationGroupDeleted(_ groupId: String) {
    notifyEvent(serviceName(), "onConversationGroupDeleted", ["groupId": groupId])
  }

  /**
   *  会话分组变更回调
   *
   *  @param conversationGroup 会话分组
   */
  func onConversationGroupChanged(_ conversationGroup: V2NIMConversationGroup) {
    notifyEvent(serviceName(), "onConversationGroupChanged", conversationGroup.toDic())
  }

  /**
   *  会话分组删除回调
   *
   *  @param groupId 分组id
   *  @param conversations 会话列表
   */
  func onConversationsAdded(toGroup groupId: String, conversations: [V2NIMConversation]) {
    let conversationsDic = conversations.map { $0.toDictionary() }
    notifyEvent(serviceName(), "onConversationsAddedToGroup", ["groupId": groupId,
                                                               "conversations": conversationsDic])
  }

  /**
   *  会话分组删除回调
   *
   *  @param groupId 分组id
   *  @param conversationIds 会话id列表
   */
  func onConversationsRemoved(fromGroup groupId: String, conversationIds: [String]) {
    notifyEvent(serviceName(), "onConversationsRemovedFromGroup", ["groupId": groupId,
                                                                   "conversationIds": conversationIds])
  }
}

File: nim_core_v2/android/src/main/kotlin/com/netease/nimflutter/services/FLTConversationGroupService.kt
/*
 * Copyright (c) 2022 NetEase, Inc. All rights reserved.
 * Use of this source code is governed by a MIT license that can be
 * found in the LICENSE file.
 */

package com.netease.nimflutter.services

import android.content.Context
import com.netease.nimflutter.FLTConstant
import com.netease.nimflutter.FLTService
import com.netease.nimflutter.NimCore
import com.netease.nimflutter.NimResult
import com.netease.nimflutter.extension.toMap
import com.netease.nimlib.sdk.NIMClient
import com.netease.nimlib.sdk.v2.conversation.V2NIMConversationGroupListener
import com.netease.nimlib.sdk.v2.conversation.V2NIMConversationGroupService
import com.netease.nimlib.sdk.v2.conversation.model.V2NIMConversation
import com.netease.nimlib.sdk.v2.conversation.model.V2NIMConversationGroup
import com.netease.yunxin.kit.alog.ALog
import kotlin.coroutines.resume
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.channels.onFailure
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.suspendCancellableCoroutine

class FLTConversationGroupService(
    applicationContext: Context,
    nimCore: NimCore
) : FLTService(applicationContext, nimCore) {
    private val tag = "FLTConversationGroupService"
    override val serviceName = "V2NIMConversationGroupService"

    init {
        nimCore.onInitialized {
            conversationGroupListener()
            registerFlutterMethodCalls(
                "createConversationGroup" to ::createConversationGroup,
                "deleteConversationGroup" to ::deleteConversationGroup,
                "updateConversationGroup" to ::updateConversationGroup,
                "addConversationsToGroup" to ::addConversationsToGroup,
                "removeConversationsFromGroup" to ::removeConversationsFromGroup,
                "getConversationGroup" to ::getConversationGroup,
                "getConversationGroupList" to ::getConversationGroupList,
                "getConversationGroupListByIds" to ::getConversationGroupListByIds
            )
        }
    }

    @ExperimentalCoroutinesApi
    private fun conversationGroupListener() {
        callbackFlow<Pair<String, Map<String, Any?>?>> {
            val listener = object : V2NIMConversationGroupListener {
                override fun onConversationGroupCreated(conversationGroup: V2NIMConversationGroup?) {
                    ALog.i(serviceName, "onConversationGroupCreated: ${conversationGroup?.name}")
                    trySend(
                        Pair<String, Map<String, Any?>?>(
                            "onConversationGroupCreated",
                            conversationGroup?.toMap()
                        )
                    ).onFailure {
                        ALog.i(serviceName, "send onConversationGroupCreated fail: ${it?.message}")
                    }
                }

                override fun onConversationGroupDeleted(groupId: String?) {
                    ALog.i(serviceName, "onConversationGroupDeleted: $groupId")
                    trySend(
                        Pair<String, Map<String, Any?>?>(
                            "onConversationGroupDeleted",
                            mapOf(
                                "groupId" to groupId
                            )
                        )
                    ).onFailure {
                        ALog.i(serviceName, "send onConversationGroupDeleted fail: ${it?.message}")
                    }
                }

                override fun onConversationGroupChanged(conversationGroup: V2NIMConversationGroup?) {
                    ALog.i(serviceName, "onConversationGroupChanged: ${conversationGroup?.name}")
                    trySend(
                        Pair<String, Map<String, Any?>?>(
                            "onConversationGroupChanged",
                            conversationGroup?.toMap()
                        )
                    ).onFailure {
                        ALog.i(serviceName, "send onConversationGroupChanged fail: ${it?.message}")
                    }
                }

                override fun onConversationsAddedToGroup(groupId: String?, conversations: MutableList<V2NIMConversation>?) {
                    ALog.i(serviceName, "onConversationsAddedToGroup: $conversations")
                    trySend(
                        Pair<String, Map<String, Any?>?>(
                            "onConversationsAddedToGroup",
                            mapOf(
                                "groupId" to groupId,
                                "conversations" to conversations?.map { it.toMap() }
                            )
                        )
                    ).onFailure {
                        ALog.i(serviceName, "send onConversationsAddedToGroup fail: ${it?.message}")
                    }
                }

                override fun onConversationsRemovedFromGroup(groupId: String?, conversationIds: MutableList<String>?) {
                    ALog.i(serviceName, "onConversationsRemovedFromGroup: $conversationIds")
                    trySend(
                        Pair<String, Map<String, Any?>?>(
                            "onConversationsRemovedFromGroup",
                            mapOf(
                                "groupId" to groupId,
                                "conversationIds" to conversationIds
                            )
                        )
                    ).onFailure {
                        ALog.i(serviceName, "send onConversationsRemovedFromGroup fail: ${it?.message}")
                    }
                }
            }

            NIMClient.getService(V2NIMConversationGroupService::class.java).apply {
                this.addConversationGroupListener(listener)
                awaitClose {
                    this.removeConversationGroupListener(listener)
                }
            }
        }.onEach { event ->
            notifyEvent(
                method = event.first,
                arguments = event.second as Map<String, Any?>
            )
        }.launchIn(nimCore.lifeCycleScope)
    }

    private suspend fun createConversationGroup(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val name = arguments["name"] as? String
            val serverExtension = arguments["serverExtension"] as? String
            val conversationIds = arguments["conversationIds"] as? List<String>

            if (name == null) {
                cont.resume(NimResult(code = FLTConstant.paramErrorCode, errorDetails = "createConversationGroup name is empty!"))
            } else {
                NIMClient.getService(V2NIMConversationGroupService::class.java)
                    .createConversationGroup(
                        name,
                        serverExtension,
                        conversationIds,
                        { result ->
                            cont.resume(
                                NimResult(
                                    code = 0,
                                    data = result.toMap()
                                )
                            )
                        },
                        { error ->
                            cont.resume(
                                if (error != null) {
                                    NimResult(code = error.code, errorDetails = error.desc)
                                } else {
                                    NimResult(code = -1, errorDetails = "createConversationGroup failed!")
                                }
                            )
                        }
                    )
            }
        }
    }

    private suspend fun deleteConversationGroup(arguments: Map<String, *>): NimResult<Nothing> {
        return suspendCancellableCoroutine { cont ->
            val groupId = arguments["groupId"] as? String

            if (groupId == null) {
                cont.resume(NimResult(code = FLTConstant.paramErrorCode, errorDetails = "deleteConversationGroup groupId is empty!"))
            } else {
                NIMClient.getService(V2NIMConversationGroupService::class.java)
                    .deleteConversationGroup(
                        groupId,
                        {
                            cont.resume(
                                NimResult.SUCCESS
                            )
                        },
                        { error ->
                            cont.resume(
                                if (error != null) {
                                    NimResult(code = error.code, errorDetails = error.desc)
                                } else {
                                    NimResult(code = -1, errorDetails = "deleteConversationGroup failed!")
                                }
                            )
                        }
                    )
            }
        }
    }

    private suspend fun updateConversationGroup(arguments: Map<String, *>): NimResult<Nothing> {
        return suspendCancellableCoroutine { cont ->
            val groupId = arguments["groupId"] as? String
            val name = arguments["name"] as? String
            val serverExtension = arguments["serverExtension"] as? String
            if (groupId == null) {
                cont.resume(NimResult(code = FLTConstant.paramErrorCode, errorDetails = "updateConversationGroup groupId is empty!"))
            } else {
                NIMClient.getService(V2NIMConversationGroupService::class.java)
                    .updateConversationGroup(
                        groupId,
                        name,
                        serverExtension,
                        {
                            cont.resume(
                                NimResult.SUCCESS
                            )
                        },
                        { error ->
                            cont.resume(
                                if (error != null) {
                                    NimResult(code = error.code, errorDetails = error.desc)
                                } else {
                                    NimResult(code = -1, errorDetails = "createConversationGroup failed!")
                                }
                            )
                        }
                    )
            }
        }
    }

    private suspend fun addConversationsToGroup(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val groupId = arguments["groupId"] as? String
            val conversationIds = arguments["conversationIds"] as? List<String>

            if (groupId == null || conversationIds == null) {
                cont.resume(NimResult(code = FLTConstant.paramErrorCode, errorDetails = "addConversationsToGroup name or conversationIds is empty!"))
            } else {
                NIMClient.getService(V2NIMConversationGroupService::class.java)
                    .addConversationsToGroup(
                        groupId,
                        conversationIds,
                        { result ->
                            cont.resume(
                                NimResult(
                                    code = 0,
                                    data =
                                    mutableMapOf(
                                        "conversationOperationResults" to
                                            result.map { it.toMap() }
                                    )
                                )
                            )
                        },
                        { error ->
                            cont.resume(
                                if (error != null) {
                                    NimResult(code = error.code, errorDetails = error.desc)
                                } else {
                                    NimResult(code = -1, errorDetails = "addConversationsToGroup failed!")
                                }
                            )
                        }
                    )
            }
        }
    }

    private suspend fun removeConversationsFromGroup(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val groupId = arguments["groupId"] as? String
            val conversationIds = arguments["conversationIds"] as? List<String>

            if (groupId == null || conversationIds == null) {
                cont.resume(NimResult(code = FLTConstant.paramErrorCode, errorDetails = "removeConversationsFromGroup name or conversationIds is empty!"))
            } else {
                NIMClient.getService(V2NIMConversationGroupService::class.java)
                    .removeConversationsFromGroup(
                        groupId,
                        conversationIds,
                        { result ->
                            cont.resume(
                                NimResult(
                                    code = 0,
                                    data =
                                    mutableMapOf(
                                        "conversationOperationResults" to
                                            result.map { it.toMap() }
                                    )
                                )
                            )
                        },
                        { error ->
                            cont.resume(
                                if (error != null) {
                                    NimResult(code = error.code, errorDetails = error.desc)
                                } else {
                                    NimResult(code = -1, errorDetails = "removeConversationsFromGroup failed!")
                                }
                            )
                        }
                    )
            }
        }
    }

    private suspend fun getConversationGroup(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val groupId = arguments["groupId"] as? String

            if (groupId == null) {
                cont.resume(NimResult(code = FLTConstant.paramErrorCode, errorDetails = "getConversationGroup name is empty!"))
            } else {
                NIMClient.getService(V2NIMConversationGroupService::class.java)
                    .getConversationGroup(
                        groupId,
                        { result ->
                            cont.resume(
                                NimResult(
                                    code = 0,
                                    data = result.toMap()
                                )
                            )
                        },
                        { error ->
                            cont.resume(
                                if (error != null) {
                                    NimResult(code = error.code, errorDetails = error.desc)
                                } else {
                                    NimResult(code = -1, errorDetails = "getConversationGroup failed!")
                                }
                            )
                        }
                    )
            }
        }
    }

    private suspend fun getConversationGroupList(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            NIMClient.getService(V2NIMConversationGroupService::class.java)
                .getConversationGroupList(
                    { result ->
                        cont.resume(
                            NimResult(
                                code = 0,
                                data =
                                mutableMapOf(
                                    "conversationGroups" to result.map { it.toMap() }
                                )
                            )
                        )
                    },
                    { error ->
                        cont.resume(
                            if (error != null) {
                                NimResult(code = error.code, errorDetails = error.desc)
                            } else {
                                NimResult(code = -1, errorDetails = "getConversationGroupList failed!")
                            }
                        )
                    }
                )
        }
    }

    private suspend fun getConversationGroupListByIds(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->

            val groupIds = arguments["groupIds"] as? List<String>

            if (groupIds == null) {
                cont.resume(NimResult(code = FLTConstant.paramErrorCode, errorDetails = "getConversationGroupListByIds groupIds is empty!"))
            } else {
                NIMClient.getService(V2NIMConversationGroupService::class.java)
                    .getConversationGroupListByIds(
                        groupIds,
                        { result ->
                            cont.resume(
                                NimResult(
                                    code = 0,
                                    data =
                                    mutableMapOf(
                                        "conversationGroups" to result.map { it.toMap() }
                                    )
                                )
                            )
                        },
                        { error ->
                            cont.resume(
                                if (error != null) {
                                    NimResult(code = error.code, errorDetails = error.desc)
                                } else {
                                    NimResult(code = -1, errorDetails = "getConversationGroupListByIds failed!")
                                }
                            )
                        }
                    )
            }
        }
    }
}
